drivers/i2c/i2c_mspm0.c

```diff
@@ -0,0 +1,814 @@
+	uint32_t clock_rate;
+	uint32_t tick_cycles;
+	uint64_t timeout_cycles;
+	uint32_t ticks_needed;
+	uint16_t counter_value;
+	int ret;
+
+	ret = clock_control_get_rate(clk_dev, (struct mspm0_sys_clock *)config->clock_subsys,
+				     &clock_rate);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Each count is equal to (1 + TPR) * 12 functional clocks */
+	tick_cycles = (period + 1) * 12;
+	timeout_cycles = (uint64_t)timeout_ms * (clock_rate / 1000);
+	ticks_needed = (timeout_cycles + tick_cycles - 1) / tick_cycles;
+	/* Lower 4-bits of counter are automatically set to 0x0 */
+	counter_value = ticks_needed >> 4;
+	if (counter_value > 0xFF) {
+		return -EINVAL;
+	}
+
+	DL_I2C_enableTimeoutA(config->base);
+	DL_I2C_setTimeoutACount(config->base, counter_value);
+	return 0;
+}
+#endif /* CONFIG_I2C_SCL_LOW_TIMEOUT != 0 */
+
+static int i2c_mspm0_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+	const struct device *const clk_dev = DEVICE_DT_GET(DT_NODELABEL(ckm));
+	uint32_t clock_rate;
+	uint32_t desired_speed;
+	int32_t period;
+	int ret;
+
+	ret = clock_control_get_rate(clk_dev, (clock_control_subsys_t)config->clock_subsys,
+				     &clock_rate);
+	if (ret < 0) {
+		ret = -ENODEV;
+		goto out;
+	}
+
+	if (dev_config & I2C_MSG_ADDR_10_BITS) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	k_sem_take(data->i2c_busy_sem, K_FOREVER);
+
+	/* Config I2C speed */
+	switch (I2C_SPEED_GET(dev_config)) {
+	case I2C_SPEED_STANDARD:
+		desired_speed = 100000;
+		break;
+	case I2C_SPEED_FAST:
+		desired_speed = 400000;
+		break;
+	default:
+		ret = -EINVAL;
+		goto sem_give;
+	}
+
+	/* Calculate the timer period based on the desired speed and clock rate.
+	 * This function works out to be ceil(clockRate/(desiredSpeed*10)) - 1
+	 */
+	period = clock_rate / (desired_speed * 10);
+	period -= (clock_rate % (desired_speed * 10) == 0) ? 1 : 0;
+
+	/* Set the I2C speed */
+	if (period <= 0) {
+		ret = -EINVAL;
+		goto sem_give;
+	}
+
+	DL_I2C_setTimerPeriod(config->base, period);
+	data->dev_config = dev_config;
+
+#if (CONFIG_I2C_SCL_LOW_TIMEOUT != 0)
+	ret = i2c_mspm0_configure_timeout(dev, period, CONFIG_I2C_SCL_LOW_TIMEOUT);
+#endif /* CONFIG_I2C_SCL_LOW_TIMEOUT */
+
+	/* Config other settings */
+	DL_I2C_setControllerTXFIFOThreshold(config->base, DL_I2C_TX_FIFO_LEVEL_BYTES_1);
+	DL_I2C_setControllerRXFIFOThreshold(config->base, DL_I2C_RX_FIFO_LEVEL_BYTES_1);
+	DL_I2C_enableControllerClockStretching(config->base);
+
+	/* Configure Interrupts */
+	DL_I2C_enableInterrupt(config->base, TI_MSPM0_CONTROLLER_INTERRUPTS);
+
+	/* Enable module */
+	DL_I2C_enableController(config->base);
+
+sem_give:
+	k_sem_give(data->i2c_busy_sem);
+out:
+	return ret;
+}
+
+static int i2c_mspm0_get_config(const struct device *dev, uint32_t *dev_config)
+{
+	struct i2c_mspm0_data *data = dev->data;
+	*dev_config = data->dev_config;
+	return 0;
```

Comment:
if the i2c is not configured, this API shall return error as per API description